### PR TITLE
update mpAndroidChart to V3

### PR DESCRIPTION
**Pull request by wuxudong**
update mpAndroidChart to V3

add several new features. like xValues as number, custom marker text.

### CODE REVIEW

**Review by mskec — CHANGES_REQUESTED** [review]
@wuxudong thank you for PR!
Please move example to example repo.

**Comment by wuxudong** [comment]
I revert the examples. 

A PR of examples is coming soon.

**Comment by wuxudong** [comment]
Is there anything more I should change?

**Comment by mskec** [comment]
I need to check if there are breaking changes and test on example app.
I'm quite busy atm, but I will try to do it in the next few days.

File: android/src/main/java/com/github/reactNativeMPAndroidChart/charts/ChartBaseManager.java [hunk 1]
```diff
@@ -103,7 +107,14 @@ public void setLegend(T chart, ReadableMap propMap) {
                         colorsParsed[i] = Color.parseColor(colors[i]);
                     }
 
-                    legend.setCustom(colorsParsed, labels);
+                    LegendEntry[] legendEntries = new LegendEntry[labels.length];
+                    for (int i = 0; i < legendEntries.length; i++) {
```
Comment (mskec) [inline android/src/main/java/com/github/reactNativeMPAndroidChart/charts/ChartBaseManager.java]: reuse for loop where colors are parsed

File: README.md [hunk 2]
```diff
@@ -1,7 +1,7 @@
 # React Native MPAndroidChart
 This library is React Native wrapper of popular Android charting library [MPAndroidChart](https://github.com/PhilJay/MPAndroidChart).
 
-
+ update version to v3.0.1
```
Comment (mskec) [inline README.md]: rm

File: android/src/main/java/com/github/reactNativeMPAndroidChart/charts/LineChartManager.java [hunk 3]
```diff
@@ -92,4 +95,21 @@ void dataSetConfig(IDataSet<Entry> dataSet, ReadableMap config) {
             lineDataSet.enableDashedLine(lineLength, spaceLength, phase);
         }
     }
+
+    @Override
+    Entry createEntry(ReadableArray values, int index) {
```
Comment (mskec) [inline android/src/main/java/com/github/reactNativeMPAndroidChart/charts/LineChartManager.java:101]: can you set this as default implementation in ChartBaseManager?

File: android/src/main/java/com/github/reactNativeMPAndroidChart/charts/YAxisChartBase.java [hunk 4]
```diff
@@ -21,10 +21,10 @@
 
     protected void setYAxisConfig(YAxis axis, ReadableMap propMap) {
         if (BridgeUtils.validate(propMap, ReadableType.Number, "axisMaxValue")) {
-            axis.setAxisMaxValue((float) propMap.getDouble("axisMaxValue"));
+            axis.setAxisMaximum((float) propMap.getDouble("axisMaxValue"));
```
Comment (mskec) [inline android/src/main/java/com/github/reactNativeMPAndroidChart/charts/YAxisChartBase.java]: Update prop name to match new method name. `axisMaxValue` -> `axisMaximum`.
Please do that for all method changes.

Additionally we could keep old names alongs side new and add deprecation warning if old is used.

File: .gitignore [hunk 5]
```diff
@@ -112,3 +112,9 @@ crashlytics.properties
 crashlytics-build.properties
 fabric.properties
 
+
+.idea
```
Comment (mskec) [inline .gitignore]: please remove changes you made because of examples app in this repo

File: lib/PieChart.js [hunk 6]
```diff
@@ -26,7 +26,10 @@ const iface = {
     // TODO PieChart should have only one dataset
     data: PropTypes.shape({
       datasets: PropTypes.arrayOf(PropTypes.shape({
-        yValues: PropTypes.arrayOf(PropTypes.number),
+        values: PropTypes.arrayOf(PropTypes.shape({
+          y: PropTypes.number.isRequired,
```
Comment (mskec) [inline lib/PieChart.js]: why not `value` instead of `y` for this chart? Please also change that for Radar chart

**Review by mskec — CHANGES_REQUESTED** [review]
I have taken a look at what you have done and it’s mostly working fine, however there are some breaking changes that I am not sure how to proceed with.

My goal is to have as little as possible of breaking changes and to have clear migration guide for existing users.

Could you make a list of all API changes? You can write them here and I will include them in release notes.

Regarding [values in datasets](https://github.com/mskec/react-native-mp-android-chart-example/pull/5/files#diff-d4b0988aa897088a8fde2b9f3d8c80b9R30). What do you think about allowing existing way of use as array of numbers for charts where only y component is required? Here I do not mean to replace the new way, just to also allow the old way so that we do not have breaking change there.

**Comment by wuxudong** [comment]
change log in brief


* new feature:

update MpAndroidChart to V3
enable custom marker value 
add several config: highlightEnabled




* api change:

     * dataset renamed to dataSet

     * yValues renamed to values,  and format changed. 
      
         for example:
         
  
lineChart    

	yValues: [y1,y2,y3]   

->
 
	values:     
		[y1, y2, y3]
		[{y:y1}, {y:y2}, {y:y3}]
		[{x:x1, y:y1}, {x:x2, y:y2}, {x:x3, y:y3}]




pieChart

  	yValues: [y1,y2,y3]

-> 
	
       values:
		[y1, y2, y3]
		[{value:y1}, {value:y2}, {value:y3}]
		[{value:y1, label: label1}, {value:y2,  label: label2}, {value:y3,  label: label3}]

other chart's dataSet.values please refer to corresponding js.

  *  xValues in datasets are moved to xAxis.valueFormatter

 valueFormatter now support:

		"largeValue" / "percent"
		custom string format like "###.###.###"
		[xLabel1, XLable2, XLable3, XLable4,...]   // this is equal to former xValues.

**Comment by wuxudong** [comment]
code has already changed. please take a look.

File: android/src/main/java/com/github/reactNativeMPAndroidChart/charts/LineChartManager.java (hunk 3 above)
Comment (wuxudong) [reply android/src/main/java/com/github/reactNativeMPAndroidChart/charts/LineChartManager.java:101]: only LineChartManager and scatterChartManager share the default implementation. so I extract all createEntry to standalone DataSetUtils.

**Comment by wuxudong** [comment]
can work with react-native-mp-android-chart-example which is also under review.

File: android/src/main/java/com/github/reactNativeMPAndroidChart/utils/DataSetUtils.java [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+import com.facebook.react.bridge.ReadableMap;
+import com.facebook.react.bridge.ReadableType;
+import com.github.mikephil.charting.data.BarEntry;
+import com.github.mikephil.charting.data.BubbleEntry;
+import com.github.mikephil.charting.data.CandleEntry;
+import com.github.mikephil.charting.data.Entry;
+import com.github.mikephil.charting.data.PieEntry;
+import com.github.mikephil.charting.data.RadarEntry;
+
+/**
```
Comment (mskec) [inline android/src/main/java/com/github/reactNativeMPAndroidChart/utils/DataSetUtils.java:13]: Please remove this

File: lib/BubbleChart.js [hunk 8]
```diff
@@ -13,13 +13,13 @@ const iface = {
     ...ChartBase.propTypes,
 
     data: PropTypes.shape({
-      datasets: PropTypes.arrayOf(PropTypes.shape({
+      dataSets: PropTypes.arrayOf(PropTypes.shape({
         values: PropTypes.arrayOf(PropTypes.shape({
           x: PropTypes.number,
           y: PropTypes.number.isRequired,
           size: PropTypes.number.isRequired,
+          marker:PropTypes.string,
```
Comment (mskec) [inline lib/BubbleChart.js]: why did you add marker here? I do not see it being used when creating entries. Am I missing something?

File: android/src/main/java/com/github/reactNativeMPAndroidChart/charts/LineChartManager.java [hunk 9]
```diff
@@ -55,8 +59,8 @@ void dataSetConfig(IDataSet<Entry> dataSet, ReadableMap config) {
         if (BridgeUtils.validate(config, ReadableType.Boolean, "drawCircles")) {
             lineDataSet.setDrawCircles(config.getBoolean("drawCircles"));
         }
-        if (BridgeUtils.validate(config, ReadableType.Boolean, "drawCubic")) {
-            lineDataSet.setDrawCubic(config.getBoolean("drawCubic"));
+        if (BridgeUtils.validate(config, ReadableType.String, "mode")) {
```
Comment (mskec) [inline android/src/main/java/com/github/reactNativeMPAndroidChart/charts/LineChartManager.java:62]: you did not change `drawCubic` to `mode` prop type in some `lib/*Chart.js` files.

**Review by mskec — COMMENTED** [review]
Thank you for updating.
Please be more systematic when changing api. We shouldn't leave anything behind.
Would you mind if I added few commits to your PRs to speed up review?

File: lib/BubbleChart.js (hunk 8 above)
Comment (wuxudong) [reply lib/BubbleChart.js]: just like lineChart, if you add marker attribute here, then the content will displayed in marker when you select the entry.

File: android/src/main/java/com/github/reactNativeMPAndroidChart/charts/LineChartManager.java (hunk 9 above)
Comment (wuxudong) [reply android/src/main/java/com/github/reactNativeMPAndroidChart/charts/LineChartManager.java:62]: I think only LineChart have mode property.

File: android/src/main/java/com/github/reactNativeMPAndroidChart/utils/DataSetUtils.java (hunk 7 above)
Comment (wuxudong) [reply android/src/main/java/com/github/reactNativeMPAndroidChart/utils/DataSetUtils.java:13]: ? they are used in DataSetUtils->createPieEntry, createRadarEntry

Comment (wuxudong) [reply android/src/main/java/com/github/reactNativeMPAndroidChart/utils/DataSetUtils.java:13]: Do you mean remove the whole class?

File: lib/BubbleChart.js (hunk 8 above)
Comment (wuxudong) [reply lib/BubbleChart.js]: every chart type can have marker property in their data entry.

File: android/src/main/java/com/github/reactNativeMPAndroidChart/utils/DataSetUtils.java (hunk 7 above)
Comment (mskec) [reply android/src/main/java/com/github/reactNativeMPAndroidChart/utils/DataSetUtils.java:13]: No I meant remove comment

File: android/src/main/java/com/github/reactNativeMPAndroidChart/charts/LineChartManager.java (hunk 9 above)
Comment (mskec) [reply android/src/main/java/com/github/reactNativeMPAndroidChart/charts/LineChartManager.java:62]: oh sorry, you are right

**Review by mskec — APPROVED** [review]
this is ok, I will merge it to `v3` branch first

**Pull request merged**